### DISCUSSION TITLE
For `@mdx-js/react`, move `@types/react` from `dependencies` to `peerDependencies`

### DISCUSSION THREAD

**Discussion by unrevised6419** [mdx-js/mdx#2238] [General]
In the `@mdx-js/reaact` is mentioned this:

> To enable types for imported .mdx, .md, etcetera files, you should make sure the TypeScript JSX namespace is typed. This is done by installing and using the types of your framework, such as [@types/react](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/HEAD/types/react).
> – https://github.com/mdx-js/mdx/tree/main/packages/react#types

So it says that the developer itself should install the JSX types he needs, `@types/react` or maybe others like `@vue/runtime-dom`

But there is a problem, `@types/react` [is included](https://github.com/mdx-js/mdx/blob/f4d78be5c015e8fab48589efae9507d0304e5e94/packages/react/package.json#L45) in package `dependencies`, and is creating conflicts with other JSX types, like `@vue/runtime-dom`, because TypeScript prefers as default `@types/react`, if it will find it in `node_modules`

From where this discussion was born?

When having a Vue/Nuxt project, it recommends using `Volar` extensions that will deal great with TypeScript and adds `@vue/runtime-dom` JSX types by default implicitly.

At the moment you want to integrate Storybook, the problem appears, because it will add `@types/react` as a transitive dependency from `@mdx-js/react`

```shell
npm ls @types/react
the-project@0.0.0 /Users/iamandrewluca/Projects/the-project
└─┬ @storybook/addon-essentials@7.0.0-beta.32
  └─┬ @storybook/addon-docs@7.0.0-beta.32
    └─┬ @mdx-js/react@2.2.1
      └── @types/react@18.0.27
```

Now TypeScript will see `@types/react` in `node_modules` and all your JSX in Vue will be Typed as React.

![image](https://user-images.githubusercontent.com/1881266/214836318-3feb7e35-e1aa-4604-a632-180304b82166.png)

The [current solution](https://github.com/johnsoncodehk/volar/discussions/592#discussioncomment-1763880) is to do some hacks to stub `@types/react` to an empty package

```shell
npm ls @types/react
my-project@0.0.0 /Users/iamandrewluca/Projects/my-project
├─┬ @storybook/addon-essentials@7.0.0-beta.32
│ └─┬ @storybook/addon-docs@7.0.0-beta.32
│   └─┬ @mdx-js/react@2.2.1
│     └── @types/react@18.0.27
└── @types/react@0.0.0 -> ./stub/types__react
```

Now TypeScript will see that there is not JSX namespace in `@types/react` and will go back to using `@vue/runtime-dom`

![image](https://user-images.githubusercontent.com/1881266/214836785-6468ea4b-1e93-429a-86bd-11e52b11adb7.png)

The suggestion from our side would be to move `@types/react` from `dependencies` to `devDependencies` so it wont become a transitive dependency.

This was discussed in depth in some threads:
- https://github.com/johnsoncodehk/volar/discussions/592
- https://github.com/storybookjs/storybook/issues/12505

**Comment by wooorm** [comment]
Hey!

`@types/react` is used in `@mdx-js/react`. It’s an actual dependency. It’s needed: 
https://github.com/mdx-js/mdx/blob/f4d78be5c015e8fab48589efae9507d0304e5e94/packages/react/lib/index.js#L2

Without it, this project can’t be used in TypeScript.

**Reply by unrevised6419** [reply]
When someone is using `@mdx-js/react` with TypeScript, that means they already have installed `react` and `@types/react` explicitly in their project.

**Reply by unrevised6419** [reply]
If you mean that `@types/react` is used for the development of `@mdx-js/react` package, moving `@types/react` to `devDependencies` should not break anything.

**Reply by wooorm** [reply]
So likely storybook will also have it as a dependency?

And, same with react. Which is why it’s included in our (peer) deps too

**Reply by wooorm** [reply]
No. It’s in the actual types we ship.

**Reply by unrevised6419** [reply]
Yes, Storybook actually also had it as a `dependency` and moved it to `devDependencies` because it created problems in userland.

The final consumer will add `@types/react` or any other JSX types

**Reply by unrevised6419** [reply]
If any dependents do not have `@types/react` installed explicitly, this would be a breaking change for them.

We need to find some `@mdx-js/react` [dependants](https://www.npmjs.com/package/@mdx-js/react?activeTab=dependents) that use TypeSCript and see if they have it explicitly installed

**Reply by remcohaszing** [reply]
Since `react` is a peer dependency, I believe `@types/react` should also be a peer dependency. However, since non-TypeScript users shouldn't be bothered with it, it should be an optional peer dependency.

Having it a regular dependency causes issues for React 16 and 17 users. They now need to resort to npm overrides / yarn resolutions.

The problem reported here is an edge case, but this would solve it by coincidence.

**Reply by unrevised6419** [reply]
Starting with `npm@7` peer dependencies will also be installed, it is possible that it will still create conflicts 🤔

**Reply by wooorm** [reply]
> it should be an optional peer dependency.

We’ve been adding types packages to `dependencies` *for years*. Why are you suddenly suggesting to change to optional dependencies?

> Having it a regular dependency causes issues for React 16 and 17 users

Where does it do that? Can you please reproduce?

> but this would solve it by coincidence.

What would it solve?

**Reply by remcohaszing** [reply]
@iamandrewluca 

It will install peer dependencies, but it will make sure the dependency chain is flat. It will not install optional peer dependencies. This means it would install the `react` version specified by the user, but use `react>=16` is nothing is specified. It would not install `@types/react`, but it will make sure a matching version is used if the user installs it.

Note that if `react` shipped their own types, it would still be installed and this issue wouldn't be solved. You can also use the `types` compiler option in `tsconfig.json` to cherry pick the types that TypeScript should load.

**Reply by wooorm** [reply]
What is `it`? Which point are you replying to? Sorry, but I don’t understand

**Reply by remcohaszing** [reply]
Sorry, I was slow to respond, GitHub was slow to update the thread for me.

> We’ve been adding types packages to `dependencies` for *years.* Why are you suddenly suggesting to change to optional dependencies?

I’m not suggesting to stop adding types packages to `dependencies`. I’m suggesting that:

- Just like imports in runtime code, types packages used in the built `.d.ts` files should be dependencies.
- Framework dependencies like `react` / `preact` / `vue` / `angular` should be peer dependencies.
  - Types dependencies for such frameworks should also be peer dependencies.
  - Types peer dependencies should be optional, because it shouldn’t bother non-TypeScript users.

In the case of `@mdx-js/react`, `react` is correctly a `peerDependency`. Because `react` is a `peerDependency`, so should `@types/react` be.

**Reply by remcohaszing** [reply]
A real world example is the following `package.json`:

```json
{
  "name": "project",
  "dependencies": {
    "@mdx-js/react": "^2.0.0",
    "@types/react": "^17.0.0",
    "react": "^17.0.0"
  }
}
```

When you place this in a directory, and run `yarn`, two versions of `@types/react` will be installed. `@mdx-js/react` will use the types for React 18, whereas the rest of the project uses types for React 17.

Interestingly yarn uses an algorithm which prefers the latest version, whereas npm appears to use an algorithm that focuses on deduplication. As a result, this issue doesn’t exist for npm users (by coincidence).

**Reply by wooorm** [reply]
- Your point 2a does not solve this user’s question. Feel free to change it, `react-markdown` also does that. I’m not opposed, I just feel it’s a bit off topic
- Your point 2b is new. Why do you want to do that, now? Why do you treat framework types different from non-framework types? There’s also `optionalDependencies`, but you don‘t suggest using that? I don’t think this is what optional (peer, dev) dependencies: https://docs.npmjs.com/cli/v9/configuring-npm/package-json#optionaldependencies. I see them as having to be handled gracefully by the code when missing. Types can’t do that.

**Reply by wooorm** [reply]
> A real world example is the following `package.json`:

Wouldn’t `peerDependencies` have the exact same problem? Because some package managers *also* install those?

**Reply by unrevised6419** [reply]
I also agree that with `peerDependencies` the problem may still exist, while with `optionalDependencies` it may work 🤔

**Reply by unrevised6419** [reply]
Reading the docs, it seems `optionalDependencies` won't solve it.

**Reply by wooorm** [reply]
No, I don’t think any of these is going to solve your use case.
And even if we’d totally drop it (which I think is a bad idea, I’m against), other tools that use React will also list React and its types.
I think your issue is more generally with TypeScript somehow picking up on a deep down dependency, which no code actually uses. And this is something that will happen to more TypeScript users. So solving it lies with TypeScript and its ecosystem.

**Reply by remcohaszing** [reply]
Point 2a on its own indeed doesn’t solve this user’s question, but it’s a prerequisite for 2b.

- `dependencies`: are always installed
- `devDependencies`: are never installed
- `optionalDependencies`: are installed if possible. This means they match the `os` and `cpu` fields in `package.json`, and their `postinstall` script doesn’t fail. If the `os` or `cpu` doesn’t match, or the `postinstall` script fails, the dependency won’t be installed, but the `npm install` command does succeed. The only use case for this I can think of, are native Node bindings.
- `peerDependencies` are dependencies that should be installed, but must be installed as a flat dependency. As of npm 8, `peerDependencies` are installed, but npm will fail if it can’t resolve them in a flat dependency tree. Using yarn or older versions of npm, it was the user’s responsibility to install these dependencies. For example:
  - ESLint / remark / Prettier / Stylelint needs to be able to resolve configurations and plugins directly from a project’s `node_modules` folder.
  - There may not be multiple versions of a framework like `react` or `angular` loaded in the same context.
- Optional `peerDependencies` are just regular peer dependencies, but the software will also work if these peer dependencies are missing.
  - For example: If a user uses TypeScript and `react`, they import from `react`. This means they are supposed to install both `react` and `@types/react`. These should both be `peerDependencies` of `@mdx-js/react`, because there should only be one occurrence of them.
  - If the user **doesn’t** use TypeScript, they shouldn’t be bothered with the `@types/react` dependency. Therefor it should be optional. In fact, they could be using `react@17`, whereas `@mdx-js/react → @types/react` would resolve to version 18, so their editor might even give invalid type hints.

So what I’m basically saying is:

- Treat `@types` dependencies the same as the package they represent (in this case `peerDependencies`)
- Don’t bother non-TypeScript users with warnings if they don’t install types dependencies (only relevant for missing `peerDependencies`, not `dependencies`)

This would solve the OP’s issue **by coincidence**. They are choosing use both Vue and React (through storybook) in the same project. If `react` did ship their own types, they would still have this issue.

**Reply by unrevised6419** [reply]
Thanks for your responses folks 🙂 will investigate further maybe we can solve it from other place 🙌🏼

**Reply by ChristianMurphy** [reply]
>  I think your issue is more generally with TypeScript somehow picking up on a deep down dependency, which no code actually uses. And this is something that will happen to more TypeScript users. So solving it lies with TypeScript and its ecosystem.

I'd agree with this point.

Where type dependencies should be is rather complex.
TypeScript considers types to be required, but some users don't want the typings or only want pieces of the typings. (ref https://github.com/syntax-tree/unist-util-is/issues/14, https://github.com/unifiedjs/unified/issues/45, https://github.com/unifiedjs/unified/pull/146, https://github.com/syntax-tree/mdast-util-gfm-table/pull/3)
TypeScript doesn't have an official/documented way of choosing which globals from typings appear.

This should really be handled at the TypeScript language level, the current approach of fiddle with which category of dependency types are labelled as isn't sustainable.
@remcohaszing noted some of the different approaches https://github.com/orgs/mdx-js/discussions/2238#discussioncomment-4788837
What he omitted is that each of the dependency managers (npm, yarn, pnpm) handles the categories of dependencies a bit different as well as handling version resolution different.
Even _within_ the _same_ dependency manager, it can handle categories and versioning different between major or minor version of itself.

There is also the related problem of sites with hybrid client side, server side, and web worker environments.
Where different files within the same project can have completely different type dependencies. 😅 (related https://github.com/unifiedjs/unified/issues/45, https://github.com/remarkjs/remark/issues/394)

---

The path forward is likely with TypeScript, there are a few threads on these topics already, notably: https://github.com/microsoft/TypeScript/pull/49886, https://github.com/microsoft/TypeScript/issues/52433, https://github.com/microsoft/TypeScript/issues/37053 which could use a signal boost and further discussion to move them forward.

**Reply by unrevised6419** [reply]
I also agree. I think I may found how to fix it in TypeScript 🤔 Will update you soon

**Reply by johnsoncodehk** [reply]
Hey folks, I have resolved a similar problem, my solution is `devDependencies` + `peerDependencies` +  `peerDependenciesMeta...optional`: https://github.com/volarjs/plugins/blob/75305d0c0c8fa571567e895d5eca54c7a446e91f/packages/typescript/package.json#L26-L33

**Reply by IanVS** [reply]
Hi, Storybook maintainer here.  We also deal with complex dependency issues in a tool that is used with and without TypeScript, and  I think ^ is the correct approach.  Regardless of whether some versions of npm have buggy behavior about auto-installing, the correct way to think about the react types is that the user should install it themselves so that they can match the version of React they are using (which means it should be a peer dependency), only some people need them (those who use TypeScript, which means using `peerDependenciesMeta.optional`), and it can also go in `devDependencies` so that it is available when developing `mdx-js/react` itself.  This would be the cleanest way to handle them, IMHO, while giving users the most control.

**Reply by zestlee1106** [reply]
Hello,

Did the above issue be resolved?

**Reply by IanVS** [reply]
No, unfortunately the types are still listed in `dependencies`: https://github.com/mdx-js/mdx/blob/f4d78be5c015e8fab48589efae9507d0304e5e94/packages/react/package.json#L45

**Reply by zestlee1106** [reply]
I see... 😭

Do you have any plans to reflect the modifications?

**Reply by wooorm** [reply]
I see no problem that can be solved here, as expression a couple of times in the thread.

a) we’re talking about `@mdx-js/react` here, a *react* specific package
b) we’re talking about a direct dependency that is used in the code. *Optional* dependencies are things that a package manager will try to install, but if it fails, it’s silent. This is needed for `os` or `platform` specific packages. Native stuff. Not for types. There’s no `optionalDependenciesForTool: {typescript: {'@types/react': '...'}}` field. Peer dependencies are for things like react which are host packages. If X has a peer dependency on Y, the user will use those tools together. That’s React for example. That’s not what happens here either.
c) even if we went with peer dependencies or optional dependencies (https://github.com/orgs/mdx-js/discussions/2238#discussioncomment-4788689), it doesn’t solve this issue, because npm will install it anyway, and all package managers do things differently (https://github.com/orgs/mdx-js/discussions/2238#discussioncomment-4789185), and presumably lots of packages will use `@types/react` as a dep or peer dep too.
d) the issue is in TypeScript: https://github.com/orgs/mdx-js/discussions/2238#discussioncomment-4789185

(Aside, storybook probably doesn’t need this package. You can pass components to `MDXContent` fine. No provider needed. Only when you want a context based API, this is useful, which you probably don’t.)

**Reply by IanVS** [reply]
>we’re talking about a direct dependency that is used in the code.

It is not used in the runtime code, only during development of the code and bundling, right?  The types compile out of the runtime, for those using JavaScript and not TypeScript.  This is why it would be added to `devDependencies` in addition to `peerDependencies`.

>even if we went with peer dependencies or optional dependencies, it doesn’t solve this issue, because npm will install it anyway,

That used to be true, but is not the case in version 9 of npm.

>presumably lots of packages will use @types/react as a dep or peer dep too.

I don't know of any other packages which use it as a direct dependency, though there might be some.  IIRC testing-library had it as a dependency for a while, and it caused a ton of problems, so they moved it.

**Reply by wooorm** [reply]
`@types/react` is used in the types we ship.
It is correct that the types are not used by people who do not use TypeScript, but we do ship them, that’s a bummer, it adds some weight.
But we ship types, for users who use TypeScript, and those types we ship need the types they depend on too.
`peerDependencies` does not solve that.

> That used to be true, but is not the case in version 9 of npm.

Do you have a reference? This is not mentioned in the changelog: https://github.blog/changelog/2022-10-24-npm-v9-0-0-released/.

**Reply by IanVS** [reply]
>Do you have a reference?

No, it was unclear to me in the docs, so I tested it out in an experiment.  I commented on [this discussion](https://github.com/npm/feedback/discussions/225#discussioncomment-4790366) as well, and someone chimed in that npm is not installing optional peerDependencies, which is also what I found in my test.

**Reply by wooorm** [reply]
Good to know.
It doesn’t sound great to me that npm made an undocumented change and that we should go all in on it.

I see that theoretically, moving frameworks to `devDependencies`, `peerDependencies`, and adding them to `peerDependenciesMeta` as `optional: true`, can be an improvement.
What makes me uneasy about this change is that our packages are operating at a humongous scale. Users will also use other package managers (yarn, pnpm), and will use outdated package managers. Some people will not have `@types/react`.
I think it’s more likely that we break peoples builds than that we improve them.

Then, this only accidentally fixes the problem at hand here (https://github.com/orgs/mdx-js/discussions/2238#discussioncomment-4788837, https://github.com/orgs/mdx-js/discussions/2238#discussioncomment-4789185).
A fix for that problem is external: it’s in TypeScript. Or in frameworks like React. Vue just landed this: https://github.com/vuejs/core/pull/7958. React should do the same.
